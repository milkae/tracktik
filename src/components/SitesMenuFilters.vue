<template>
  <div>
    Sites Menu Filters
  </div>
</template>

<script lang="ts">
import Vue, { PropOptions } from "vue";

export default Vue.extend({
  name: "SitesMenuFilters",

  props: {
    filters: {
      type: Array,
      required: true
    } as PropOptions<string[]>
  }
});
</script>
