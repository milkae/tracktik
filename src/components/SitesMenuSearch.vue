<template>
  <div>
    Sites Menu Search
  </div>
</template>

<script lang="ts">
import Vue, { PropOptions } from "vue";

export default Vue.extend({
  name: "SitesMenuSearch",

  props: {
    search: {
      type: String,
      required: true
    } as PropOptions<string>
  }
});
</script>
