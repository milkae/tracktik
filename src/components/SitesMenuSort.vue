<template>
  <div>
    Sites Menu Sort
  </div>
</template>

<script lang="ts">
import Vue, { PropOptions } from "vue";

interface Sort {
  field: string;
  order: string;
}

export default Vue.extend({
  name: "SitesMenuSort",

  props: {
    sort: {
      type: Object,
      required: true
    } as PropOptions<Sort>
  }
});
</script>
