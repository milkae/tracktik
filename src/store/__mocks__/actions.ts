import sitesFixture from "../../../tests/unit/fixtures/sites";

interface Actions {
  SEARCH_SITES: any;
}

const actions: Actions = {
  SEARCH_SITES: jest.fn().mockResolvedValue(sitesFixture)
};

export default actions;
