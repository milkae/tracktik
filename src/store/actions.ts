import { ActionTree } from "vuex";
import { RootState } from "./state";

const actions: ActionTree<RootState, RootState> = {};

export default actions;
