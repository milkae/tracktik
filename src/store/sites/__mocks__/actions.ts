import sitesFixture from "../../../../tests/unit/fixtures/sites";

interface Actions {
  SEARCH_SITES: any;
}

export const actions: Actions = {
  SEARCH_SITES: jest.fn().mockResolvedValue(sitesFixture)
};
