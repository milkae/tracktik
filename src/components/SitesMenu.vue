<template>
  <div>
    <SitesMenuFilters :filters="options.filters" />
    <SitesMenuSearch :search="options.search" />
    <SitesMenuSort :sort="options.sort" />
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import SitesMenuFilters from "@/components/SitesMenuFilters";
import SitesMenuSearch from "@/components/SitesMenuSearch";
import SitesMenuSort from "@/components/SitesMenuSort";

export default Vue.extend({
  name: "SitesMenu",

  data() {
    return {
      options: {
        filters: [],
        sort: { field: "createdAt", order: "asc" },
        search: ""
      }
    };
  },

  watch: {
    options: function(newOptions) {
      this.$emit("submitted", newOptions);
    }
  },

  components: { SitesMenuFilters, SitesMenuSearch, SitesMenuSort }
});
</script>
