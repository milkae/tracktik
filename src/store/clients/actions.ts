import { ActionTree } from "vuex";
import { RootState } from "@/store/types";
import { ClientsState } from "./types";
import api from "@/api";

export const actions: ActionTree<ClientsState, RootState> = {
  GET_CLIENTS({ commit }): any {
    return new Promise(async (resolve, reject) => {
      try {
        const clients = await api.getClients();
        commit("SET_CLIENTS", clients);
        resolve(clients);
      } catch (error) {
        reject(error);
      }
    });
  }
};
