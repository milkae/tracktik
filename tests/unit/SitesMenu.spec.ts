import { shallowMount, createLocalVue } from "@vue/test-utils";
import SitesMenu from "@/components/SitesMenu.vue";
import SitesMenuFilters from "@/components/SitesMenuFilters.vue";
import SitesMenuSort from "@/components/SitesMenuSort.vue";
import SitesMenuSearch from "@/components/SitesMenuSearch.vue";
import Vuex from "vuex";
import { state as initialState } from "@/store/sites/index";
import { actions } from "@/store/sites/actions";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("SitesMenu", () => {
  let state: any;

  const build = () => {
    const wrapper = shallowMount(SitesMenu, {
      localVue,
      store: new Vuex.Store({
        modules: {
          sites: {
            namespaced: true,
            state,
            actions
          }
        }
      })
    });

    return {
      wrapper,
      sitesMenuFilters: () => wrapper.find(SitesMenuFilters),
      sitesMenuSort: () => wrapper.find(SitesMenuSort),
      sitesMenuSearch: () => wrapper.find(SitesMenuSearch)
    };
  };

  beforeEach(() => {
    jest.resetAllMocks();
    state = { ...initialState };
  });

  it("renders the component", () => {
    const { wrapper } = build();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders main child components", () => {
    const { sitesMenuFilters, sitesMenuSort, sitesMenuSearch } = build();

    expect(sitesMenuFilters().exists()).toBe(true);
    expect(sitesMenuSort().exists()).toBe(true);
    expect(sitesMenuSearch().exists()).toBe(true);
  });

  it('calls "submitted" event when options changes', () => {
    const { wrapper } = build();
    const newOptions = {
      filters: { clientId: "1" },
      sort: { field: "createdAt", order: "asc" },
      search: "Adonis"
    };

    (wrapper.vm as any).$options.watch.options.call(wrapper.vm, newOptions);

    expect(wrapper.emitted("submitted")[0]).toEqual([newOptions]);
  });

  it("passes according binded option prop to child components", () => {
    const {
      wrapper,
      sitesMenuFilters,
      sitesMenuSort,
      sitesMenuSearch
    } = build();
    const options = {
      filters: { clientId: "1" },
      sort: { field: "createdAt", order: "asc" },
      search: "Adonis"
    };
    wrapper.setData({ options });

    expect((sitesMenuFilters().vm as any).filters).toBe(options.filters);
    expect((sitesMenuSort().vm as any).sort).toEqual(options.sort);
    expect((sitesMenuSearch().vm as any).search).toBe(options.search);
  });
});
