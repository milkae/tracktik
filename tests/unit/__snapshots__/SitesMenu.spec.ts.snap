// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SitesMenu renders the component 1`] = `
<div>
  <sitesmenufilters-stub filters=""></sitesmenufilters-stub>
  <sitesmenusearch-stub search=""></sitesmenusearch-stub>
  <sitesmenusort-stub sort="[object Object]"></sitesmenusort-stub>
</div>
`;
